Model-based Testing With Quickcheck-Dynamic

KtorZ opened the pull request.

abailly committed:
[wip] Write a simple QD model for HeadLogic

abailly committed:
[wip] Making some progress on model, now trygin to execute test

MaximilianAlgehed committed:
WIP: make StateModel play nicely with IOSim

MaximilianAlgehed committed:
WIP: fixing shrinking so that it does something (not ideal at the moment)

MaximilianAlgehed committed:
WIP: unsafeCoerce hack to get arund QuickCheck's PropertyM being a tad insane

abailly committed:
Try to get rid of some Typeable constraints but still having troubles

abailly committed:
Fix Model so that it can run in hspec :tada:

abailly committed:
Add Commits to the model

and spawns a thread to run the nodes

abailly committed:
Add Commit and Abort actions

Also distinct ctors for each action type and increase success to
increase coverage

abailly committed:
Wait for Init to perform Commit

also provide correct Eq and Show instances for Actions

KtorZ committed:
Write first assertion on the model: assert commits consistency.

  Question: how do we get a handle on the test nodes from within the property execution?

abailly committed:
Correctly compare node's state with model

Need to accumulate observed outputs

abailly committed:
Improve error message on failed assertion

KtorZ committed:
Fix recording of server outputs in TestHydraNode

KtorZ committed:
Move state-model assertion in separate function

  Now we can write moar!

abailly committed:
Assert Head is opened with the right commits

KtorZ committed:
Try labelling transitions to enrich the model output.

abailly committed:
Add more labels for transitions and generate more tests

abailly committed:
Simplify testing machinery for StateModel

abailly committed:
Unify Model to have a single consensus state

Also adapt generators to reduce the number of discarded actions

abailly committed:
Improvements to model-based test output

* Use tabulate instead of label for better classification
* Improve generators to reduce the number of discards
* Tweak frequencies to increase the number of Initial -> Open
  transitions which are more interesting

abailly committed:
Simplify Action type as a wrapper around ClientInput

Implementing actionName allows us to provide sensible names for
actions and remove the need for dedicated constructors.

abailly committed:
Reject transition by default

KtorZ committed:
Work towards generating transaction on Open

  So far, we only send to addresses owned by participants of the head. We may want to change that later, but that is tricky because, if there's no longer any utxo owned by participants, they can't spend anything.

abailly committed:
[wip] Implement perform on NewTx

We realise we need to make our model more abstract as we don't have
access to the details of the snapshots easily.

KtorZ committed:
Rework model to use abstract representation of Ledger/UTxO

  We model payments, and only within the inner-circle of the head participants. This is analogous to an account-based system. Kind of.

abailly committed:
[wip] Update generator and nextState to new tx model

KtorZ committed:
Continue plumbing mock payment model into real world.

abailly committed:
Implement missing typeclasses for Payment

This is required by IsTx but we leave them mostly undefined

KtorZ committed:
Implement concrete perform action for 'NewTx'.

KtorZ committed:
Wait for UTxO to be non-empty before performing a new transaction.

KtorZ committed:
Add safety pre-condition to 'NewTx'

  So that we do not generate newTx actions if the party has no funds.

KtorZ committed:
Add some tracing to debug infinite loop.

KtorZ committed:
Wait for transacation to be confirmed after performing NewTx.

abailly committed:
Add source of failure in CommandFailed output

KtorZ committed:
Wait for head to be open before performing new tx.

KtorZ committed:
Fix new-tx pre-condition: make sure value is correct.

  This one was a little bit tricky because it happened during shrinking. At some point, we ended up creating transactions with negative amount. Because of another failure, QC will start shrinking the sequence of commands and remove transactions from it. Thus, making past transfer non-existing and causing us to try to execute a transaction with funds we no longer own.

KtorZ committed:
Compare map of aggregated addr->values instead of list of txout

  We can have many utxo at a single address, to ordering by address like we did could still yield to inconsistency. In the end, since the model is 'account-based', we may as well treat it as such.

abailly committed:
Check equality of balances after some txs

KtorZ committed:
Capture logs of TesthydraNode in TVar.

abailly committed:
Dump node logs when ModelSpec tests errors out

abailly committed:
Adapt TUI to changed output

abailly committed:
[wip] Track snapshot production to understand Model property failure

abailly committed:
[wip] Introduce NewSn event to ensure event queue is filled

abailly committed:
Check there's always a leader for every snapshot number

abailly committed:
Filter outputs with empty values in model

Also adapt value generator

abailly committed:
Make Model similar to UTxO

abailly committed:
Improve reporting of error in case no UTxO is available

abailly committed:
Do not throw error when there's no UTxO

This is not very helpful because it actually does not count as failure
when shrinking, and it's very easy to cause when dropping an arbitrary
Tx.

abailly committed:
Dump logs captured in IOSim when property fails

abailly committed:
Custom show instance for Payment to improve error reporting

abailly committed:
Revert "[wip] Track snapshot production to understand Model property failure"

This reverts commit f38765f3ce47ec9316893f1e4fc11724c20e63c4.

abailly committed:
Start fleshing out conflict-free liveness property

abailly committed:
Ensure we don't generate 2 parties with identical keys

abailly committed:
[wip] debugging

abailly committed:
Use type alias for CardanoSigningKey

abailly committed:
Adjust wait time for UTxO to appear

* Add a delay before requying the UTxO in order to not flood node's
event queue with too many client messages and give room for other
events to be processed
* Increase time we wait for UTxO to change

abailly committed:
Move "unit" test as an example in the ModelSpec doc

ch1bo committed:
Remove empty quickcheck directory artifact (former submodule?)

abailly committed:
Fix quickcheck-dynamic package reference to remote one

abailly committed:
Rename property checking balances

abailly committed:
Rephrase liveness property

ch1bo committed:
genPayment instead of genNewTx (an Action)

This removes the Typable constraint on the model and was needed to
explore ideas on how we could use DynLogic stuff with the ActionMonad
being over IOSim

ch1bo committed:
Draft a property using the DynamicLogic

ch1bo committed:
Use a branched quickcheck-dynamic to make it compile

Still puzzling how to get it to work with (forall s. IOSim s)

abailly committed:
Make Model-based tests pass

abailly committed:
Missing golden files

abailly committed:
Clean-up and document Model

abailly committed:
Remove dynamic logic stuff

This is stashed in another branch we'll be working on soonish

ghost commented:
Fix #375

ghost commented:
@ch1bo @KtorZ If you are happy with my latest changes, I think we can merge this and move on.

ch1bo commented on hydra-node/src/Hydra/Crypto.hs:
```diff
@@ -56,6 +56,10 @@ type SignAlg = Ed25519DSIGN
 
+instance Ord SigningKey where
+  k `compare` k' =
+    serialiseSigningKeyToRawBytes k `compare` serialiseSigningKeyToRawBytes k'
```
AFAIK one should not use key material in comparison, but rather their hashes. That is at least how it's done for verification key. Maybe worth a TODO or XXX comment in the code? Not hindering merging this.

ch1bo commented on hydra-node/src/Hydra/ServerOutput.hs:
```diff
@@ -21,7 +22,7 @@ data ServerOutput tx
   | HeadIsAborted {utxo :: UTxOType tx}
   | HeadIsFinalized {utxo :: UTxOType tx}
-  | CommandFailed
+  | CommandFailed {clientInput :: ClientInput tx}
```
This should require updating the `api.yaml`! Why don't we see a test failure?

ch1bo commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -0,0 +1,579 @@
+--
+
+performs :: Monad m => Party -> ClientInput Tx -> StateT (Nodes m) m ()
+performs party command = do
```
Could: I was confusing this with the frameworks' `perform` function. Maybe `sendsInput` is a better name?

ch1bo commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -0,0 +1,579 @@
+    _ -> error $ "genPayment impossible in state: " <> show hydraState
+
+unsafeConstructorName :: (Show a) => a -> String
+unsafeConstructorName = Prelude.head . Prelude.words . show
```
We could also use https://hackage.haskell.org/package/base-4.16.1.0/docs/Data-Data.html#v:toConstr and `show` that for a safe variant of this?

ch1bo commented on hydra-node/test/Test/Util.hs:
```diff
@@ -5,36 +5,47 @@ module Test.Util where
-      $ selectTraceEventsDynamic @_ @(HydraNodeLog SimpleTx) tr
+-- | Utility function to dump logs given a `SimTrace`.
+-- TODO(SN): take a proxy instead of hard-coding HydraNodeLog
+printTrace :: forall tx a. IsTx tx => Proxy tx -> SimTrace a -> Text
```
Should: Resolve that TODO completely as this is now an exported function. Instead of the `Proxy tx`, I think we should parameterized by the trace/log type directly. i.e. `Proxy log` with `@log`

ch1bo reviewed (changes requested):
These things need addressing:
* api changed -> update api.yaml and changelog (why are we seeing no CI errors?)
* Take a `Proxy log` instead of a `Proxy tx` in `printTrace`

abailly committed:
Remove Ord instance for SigningKey

abailly committed:
Update YAML for logs following change in API

Oddly enough this did not break our tests, we probably need to
strengthen our property check here

abailly committed:
Rename performs -> sendsInput

abailly committed:
Make printTrace more generic

ch1bo commented on hydra-node/test/Test/Util.hs:
```diff
@@ -38,14 +37,13 @@ shouldRunInSim action =
 printTrace _ tr =
   unlines . map (decodeUtf8 . Aeson.encode) $
-    selectTraceEventsDynamic' @_ @(HydraNodeLog tx) tr
+    selectTraceEventsDynamic' @_ @log tr
```
:+1:

abailly committed:
Dedup parties list when generating HeadParameters

abailly committed:
Fix API schema for CommandFailed server output

ghost commented:
This is a first shot of MBT approach using Quviq's framework, I think we should merge it as-is. We have a follow-up meeting with Quviq end of the week which will provide some insights on what to do next.
